### PR TITLE
Make wasm runtime cache size configurable

### PR DESCRIPTION
**Pull request by librelois**
Polkadot companion: https://github.com/paritytech/polkadot/pull/4466

## What does it do?

Makes configurable the maximum number of different runtime versions that can be kept in cache.

## What important points should reviewers know?

Before this PR, the cache had an array of 2 `VersionedRuntime`, with a custom implementation of a "Last Recent Usage" cache strategy which consisted in swapping a runtime in 1st position when it was used.
This custom implementation is not compliant if the cache contains more than 2 elements, then a full LRU algorithm is needed, that's why I use the lru crate which does this very well.

skip check-dependent-cumulus

### CODE REVIEW

**Comment by crystalin** [comment]
@bkchr sorry to ping you, but not sure who can do a review of this one.

File: client/executor/src/wasm_runtime.rs [hunk 1]
```diff
@@ -344,7 +374,7 @@ fn decode_version(mut version: &[u8]) -> Result<RuntimeVersion, WasmError> {
 		})?
 		.into();
 
-	let core_api_id = sp_core_hashing_proc_macro::blake2b_64!(b"Core");
+	let core_api_id = sp_core::hashing::blake2_64(b"Core");
```
Comment (bkchr) [inline client/executor/src/wasm_runtime.rs:377]: Please revert

File: client/executor/src/wasm_runtime.rs [hunk 2]
```diff
@@ -54,7 +55,27 @@ impl Default for WasmExecutionMethod {
...
+	/// Wasm runtime type.
+	wasm_method: WasmExecutionMethod,
+	/// The number of WebAssembly heap pages this instance was created with.
+	heap_pages: u64,
+}
+
+struct VersionedRuntimeValue {
+	/// Shared runtime that can spawn instances.
+	module: Arc<dyn WasmModule>,
+	/// Runtime version according to `Core_version` if any.
```
Comment (bkchr) [inline client/executor/src/wasm_runtime.rs:71]:
```suggestion
	/// Runtime version if any.
```

File: client/executor/src/wasm_runtime.rs [hunk 3]
```diff
@@ -168,8 +198,16 @@ impl RuntimeCache {
 	///
 	/// `cache_path` allows to specify an optional directory where the executor can store files
 	/// for caching.
-	pub fn new(max_runtime_instances: usize, cache_path: Option<PathBuf>) -> RuntimeCache {
-		RuntimeCache { runtimes: Default::default(), max_runtime_instances, cache_path }
+	pub fn new(
```
Comment (bkchr) [inline client/executor/src/wasm_runtime.rs:201]: Docs need to be updated

File: client/executor/src/wasm_runtime.rs [hunk 4]
```diff
@@ -54,7 +55,27 @@ impl Default for WasmExecutionMethod {
...
+	/// Shared runtime that can spawn instances.
+	module: Arc<dyn WasmModule>,
+	/// Runtime version according to `Core_version` if any.
+	version: Option<RuntimeVersion>,
+	/// Cached instance pool.
+	instances: Arc<Vec<Mutex<Option<Box<dyn WasmInstance>>>>>,
+}
+
 /// A Wasm runtime object along with its cached runtime version.
+#[derive(Clone)]
```
Comment (bkchr) [inline client/executor/src/wasm_runtime.rs:78]:
```suggestion
```

File: client/executor/src/wasm_runtime.rs [hunk 5]
```diff
@@ -149,12 +179,12 @@ const MAX_RUNTIMES: usize = 2;
...
 /// request.
 ///
-/// The size of cache is equal to `MAX_RUNTIMES`.
+/// The size of cache is configurable via the cli option `--runtime-cache-size`.
 pub struct RuntimeCache {
 	/// A cache of runtimes along with metadata.
 	///
 	/// Runtimes sorted by recent usage. The most recently used is at the front.
-	runtimes: Mutex<[Option<Arc<VersionedRuntime>>; MAX_RUNTIMES]>,
+	runtimes: Mutex<LruCache<VersionedRuntimeId, VersionedRuntimeValue>>,
```
Comment (bkchr) [inline client/executor/src/wasm_runtime.rs:187]:
```suggestion
	runtimes: Mutex<LruCache<VersionedRuntimeId, VersionedRuntime>>,
```

Why don't we just use `VersionedRuntime` here, as before?

**Review by bkchr — COMMENTED** [review]
Sorry for the late review. I left some comments

File: client/executor/src/wasm_runtime.rs (hunk 5 above)
Comment (librelois) [reply client/executor/src/wasm_runtime.rs:187]: To avoid storing twice the meta-data already present in VersionedRuntimeId

**Comment by librelois** [comment]
> Sorry for the late review. I left some comments

Thanks for the review, I applied your feedback

File: client/executor/src/wasm_runtime.rs (hunk 5 above)
Comment (bkchr) [reply client/executor/src/wasm_runtime.rs:186]: Can you not remove it from the VersionedRuntime? The duplicate data? Or at least some of it? Nevertheless, nothing of that is really big, so I see no problem. Please also do this change.

Comment (librelois) [reply client/executor/src/wasm_runtime.rs]: Done

File: client/executor/src/wasm_runtime.rs [hunk 6]
```diff
@@ -54,7 +55,18 @@ impl Default for WasmExecutionMethod {
...
+	/// Runtime code hash.
+	code_hash: Vec<u8>,
+	/// Wasm runtime type.
+	wasm_method: WasmExecutionMethod,
+	/// The number of WebAssembly heap pages this instance was created with.
+	heap_pages: u64,
+}
+
 /// A Wasm runtime object along with its cached runtime version.
+#[derive(Clone)]
```
Comment (bkchr) [inline client/executor/src/wasm_runtime.rs:69]:
```suggestion
```

File: client/executor/src/wasm_runtime.rs [hunk 7]
```diff
@@ -149,12 +159,12 @@ const MAX_RUNTIMES: usize = 2;
...
 /// request.
 ///
-/// The size of cache is equal to `MAX_RUNTIMES`.
+/// The size of cache is configurable via the cli option `--runtime-cache-size`.
 pub struct RuntimeCache {
 	/// A cache of runtimes along with metadata.
 	///
 	/// Runtimes sorted by recent usage. The most recently used is at the front.
-	runtimes: Mutex<[Option<Arc<VersionedRuntime>>; MAX_RUNTIMES]>,
+	runtimes: Mutex<LruCache<VersionedRuntimeId, VersionedRuntime>>,
```
Comment (bkchr) [inline client/executor/src/wasm_runtime.rs:167]:
```suggestion
	runtimes: Mutex<LruCache<VersionedRuntimeId, Arc<VersionedRuntime>>>,
```

**Review by bkchr — COMMENTED** [review]
2 nitpicks, otherwise seems to look good.

**Comment by bkchr** [comment]
It also doesn't compile.

File: client/executor/src/wasm_runtime.rs (hunk 7 above)
Comment (librelois) [reply client/executor/src/wasm_runtime.rs]: Done

**Comment by librelois** [comment]
> It also doesn't compile.

Fixed

**Comment by bkchr** [comment]
> > It also doesn't compile.
> 
> Fixed

There are still warnings. Cargo fmt is also not happy.

**Comment by librelois** [comment]
> There are still warnings. Cargo fmt is also not happy.

Sorry about that, it should be fine now.

**Comment by bkchr** [comment]
> > There are still warnings. Cargo fmt is also not happy.
> 
> Sorry about that, it should be fine now.

I need to inform you, that CI is still not happy :see_no_evil:

**Comment by librelois** [comment]
@bkchr It seems that the CI is green now. There is no need of companion PR on polkadot or cumulus.

**Comment by bkchr** [comment]
@librelois for polkadot we need a companion. See the error. Cumulus should be fine and can be ignored.

**Review by bkchr — APPROVED** [review]

File: client/executor/src/wasm_runtime.rs [hunk 8]
```diff
@@ -149,12 +152,12 @@ const MAX_RUNTIMES: usize = 2;
 /// the memory reset to the initial memory. So, one runtime instance is reused for every fetch
 /// request.
 ///
-/// The size of cache is equal to `MAX_RUNTIMES`.
+/// The size of cache is configurable via the cli option `--runtime-cache-size`.
 pub struct RuntimeCache {
 	/// A cache of runtimes along with metadata.
 	///
 	/// Runtimes sorted by recent usage. The most recently used is at the front.
```
Comment (arkpar) [inline client/executor/src/wasm_runtime.rs:159]: Comment could use an update

**Comment by librelois** [comment]
> @librelois for polkadot we need a companion. See the error. Cumulus should be fine and can be ignored.

Can one of you take care of the polkadot companion PR or should I do it myself? And in this last case, how to proceed? (I have never made a "polkadot companion" PR).

File: client/executor/src/wasm_runtime.rs (hunk 8 above)
Comment (librelois) [reply client/executor/src/wasm_runtime.rs:159]: This comment is still true, LruCache "sorts" items by recent usage, the most recently used still at the front.

**Comment by arkpar** [comment]
The docs should probably specify that `--runtime-cache-size` is capped at 255. Btw what happens if the users sets `--runtime-cache-size=1000`?

**Review by arkpar — APPROVED** [review]

**Comment by librelois** [comment]
> The docs should probably specify that `--runtime-cache-size` is capped at 255. Btw what happens if the users sets `--runtime-cache-size=1000`?

`runtime_cache_size` is of type `u8` so it is structopt that will fail to parse the value and will print an error to the user.

**Comment by bkchr** [comment]
> > @librelois for polkadot we need a companion. See the error. Cumulus should be fine and can be ignored.
> 
> Can one of you take care of the polkadot companion PR or should I do it myself? And in this last case, how to proceed? (I have never made a "polkadot companion" PR).

You need to create a pr to polkadot that fixes the compilation with your branch. Then you need to mention it as `polkadot companion: link` in the description of this pr.

**Comment by librelois** [comment]
> You need to create a pr to polkadot that fixes the compilation with your branch. Then you need to mention it as `polkadot companion: link` in the description of this pr.

It's done

**Comment by bkchr** [comment]
@librelois can you please merge master into this branch? :see_no_evil:

**Comment by librelois** [comment]
@bkchr Looking at the log of the gitlab-check-dependent-polkadot job, there seems to be a problem in your CI: 

```
$ ./pipeline-scripts/check_dependent_project.sh paritytech substrate --substrate "$DEPENDENT_REPO" "$GITHUB_PR_TOKEN" "$CARGO_UPDATE_CRATES"
check_dependent_project
========================
This check ensures that this project's dependents do not suffer downstream breakages from new code
changes.
From https://gitlab.parity.io/parity/substrate
 * branch              master     -> FETCH_HEAD
fatal: It seems that there is already a rebase-apply directory, and
I wonder if you are in the middle of another rebase.  If that is the
case, please try
	git rebase (--continue | --abort | --skip)
If that is not the case, please
	rm -fr ".git/rebase-apply"
and run me again.  I am stopping in case you still have something
valuable there.
```

**Comment by bkchr** [comment]
bot merge

**Comment by librelois** [comment]
>  Please check https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork.

I don't have any checkboxes like "Allow edits from maintainers":

image

Maybe I need to redo the companion PR from a personal fork instead of the purestake fork.

**Comment by FlorianFranzen** [comment]
Would have been nice it that new parameter was also documented in code and that I did not have to read this issue to understand what it is for, e.g. see also [rust docs](https://paritytech.github.io/substrate/master/sc_executor/struct.WasmExecutor.html#method.new)

**Pull request merged**